lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,119 @@
+                return HttpResponse(status=426)
+
+    def process_response(self, _request, response):
+        """
+        If request is from mobile native app, then add headers to response;
+        1. EDX-APP-LATEST-VERSION; if user app version < latest available version
+        2. EDX-APP-UPGRADE-DATE; if user app version < min supported version and timestamp < deadline to upgrade
+        """
+        request_cache_dict = RequestCache.get_request_cache().data
+        upgrade_deadline = (request_cache_dict[self.UPGRADE_DEADLINE_HEADER]
+                            if self.UPGRADE_DEADLINE_HEADER in request_cache_dict else None)
+        if upgrade_deadline and upgrade_deadline != self.NO_LAST_SUPPORTED_DATE:
+            response[self.UPGRADE_DEADLINE_HEADER] = upgrade_deadline
+        latest_version = (request_cache_dict[self.LATEST_VERSION_HEADER]
+                            if self.LATEST_VERSION_HEADER in request_cache_dict else None)
+        if latest_version and latest_version != self.NO_LATEST_VERSION:
+            response[self.LATEST_VERSION_HEADER] = latest_version
+        return response
+
+    def set_request_cache(self, request, user_agent):
+        """
+        it sets request cache data for last_supported_date and latest_version with memcached values if exists against
+        user-agent else computes the values for specific platform
+        """
+        request_cache_dict = RequestCache.get_request_cache().data
+        cached_data = cache.get_many([self.UPGRADE_DEADLINE_HEADER, self.LATEST_VERSION_HEADER])
+        last_supported_date = (cached_data[self.UPGRADE_DEADLINE_HEADER]
+                               if self.UPGRADE_DEADLINE_HEADER in cached_data else None)
+        platform = self.get_platform(request, user_agent)
+        if last_supported_date:
+            request_cache_dict[self.UPGRADE_DEADLINE_HEADER] = last_supported_date
+        elif platform:
+            self.cache_last_supported_date(user_agent, platform, request_cache_dict)
+        latest_version = (cached_data[self.LATEST_VERSION_HEADER]
+                          if self.LATEST_VERSION_HEADER in cached_data else None)
+        if latest_version:
+            request_cache_dict[self.LATEST_VERSION_HEADER] = latest_version
+        elif platform:
+            self.cache_latest_version(user_agent, platform, request_cache_dict)
+        return request_cache_dict
+
+    def get_platform(self, request, user_agent):
+        """
+        determines the platform for mobile app making the request
+        returns None if request is not from native mobile app or does not belong to supported platforms
+        """
+        if is_request_from_mobile_app(request):
+            return MobilePlatform.get_instance(user_agent)
+
+    def get_cache_key_name(self, user_agent, field):
+        """ get key name to use to cache any property against user agent """
+        return "{}_{}".format(user_agent, field)
```

Comment:
The convention is to namespace key names and use ".".
For example: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/field_overrides.py#L29

So this can be "mobile_api.app_version_upgrade.last_supported_date.{useragent}".